sonar-scanner-engine/src/test/java/org/sonar/scanner/scan/filesystem/ByteCharsetDetectorTest.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * SonarQube
+ * Copyright (C) 2009-2017 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.scanner.scan.filesystem;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+import org.apache.commons.io.ByteOrderMark;
+import org.junit.Before;
+import org.junit.Test;
+
+public class ByteCharsetDetectorTest {
+  private CharsetValidation validation;
+  private ByteCharsetDetector charsets;
+
+  @Before
+  public void setUp() {
+    validation = mock(CharsetValidation.class);
+    charsets = new ByteCharsetDetector(validation, null);
+  }
+
+  @Test
+  public void detectBOM() throws URISyntaxException, IOException {
+    byte[] b = ByteOrderMark.UTF_16BE.getBytes();
+    assertThat(charsets.detectBOM(b)).isEqualTo(ByteOrderMark.UTF_16BE);
+
+    Path utf8 = Paths.get(this.getClass().getClassLoader().getResource("org/sonar/scanner/scan/filesystem/UTF-8.txt").toURI());
+    Path utf16be = Paths.get(this.getClass().getClassLoader().getResource("org/sonar/scanner/scan/filesystem/UTF-16BE.txt").toURI());
+    Path utf16le = Paths.get(this.getClass().getClassLoader().getResource("org/sonar/scanner/scan/filesystem/UTF-16LE.txt").toURI());
+    Path utf32be = Paths.get(this.getClass().getClassLoader().getResource("org/sonar/scanner/scan/filesystem/UTF-32BE.txt").toURI());
+    Path utf32le = Paths.get(this.getClass().getClassLoader().getResource("org/sonar/scanner/scan/filesystem/UTF-32LE.txt").toURI());
+
+    assertThat(charsets.detectBOM(Files.readAllBytes(utf8))).isEqualTo(ByteOrderMark.UTF_8);
```

Comment:
There could be a `getBom(String filename)` that returns `Files.readAllBytes(Paths.get(this.getClass().getClassLoader().getResource("org/sonar/scanner/scan/filesystem/" + filename).toURI()))` and reduce code duplication.